rllib/algorithms/ppo/tf/ppo_tf_rl_module.py

```diff
@@ -81,3 +83,23 @@ def _forward_train(self, batch: NestedDict):
         output[SampleBatch.ACTION_DIST_INPUTS] = action_logits
 
         return output
+
+    @override(PPORLModule)
+    def _compute_values(self, batch):
+        values = {}
+        for module_id, sa_batch in batch.policy_batches.items():
+            infos = sa_batch.pop(SampleBatch.INFOS, None)
+            sa_batch = tree.map_structure(lambda s: tf.convert_to_tensor(s), sa_batch)
+            if infos is not None:
+                sa_batch[SampleBatch.INFOS] = infos
+
+            module = self[module_id].unwrapped()
+
+            # Shared encoder.
+            encoder_outs = module.encoder(sa_batch)
+            # Value head.
+            vf_out = module.vf(encoder_outs[ENCODER_OUT][CRITIC])
```

Comment:
Also great question! :)
We are actually still doing something dumb here, which we should fix in another PR (it's unrelated to this PR and the problem occurs in other RLModules as well):

- If we have a shared value function, then everything is fine: We pass the batch through the (shared) encoder, then through only(!) the value head to get the values. Thus we don't perform any wasteful NN computations. The tensors in `encoder_outs[ENCODER_OUT][CRITIC]` and `encoder_outs[ENCODER_OUT][ACTOR]` are identical.
- However, if we have separate encoders, we pass the batch through both critic AND actor encoders to get `encoder_outs[ENCODER_OUT][CRITIC]` and `encoder_outs[ENCODER_OUT][ACTOR]`, even though, right here in this method we only need the values, NOT the policy encoder outputs. So assuming that we have a huge encoder (e.g. some large CNN stack), we would waste an additional forward pass through the policy encoder here, w/o actually requiring its outputs. This is a flaw in the API, which we should fix on a broader basis.

Reply:
To avoid this - at least for the default PPO modules - I added this `hasattr` check on the encoder object. However, this check will not prevent the superfluous policy encoder forward pass if the user has a custom encoder that does NOT have this `critic_encoder` property that we check for.
Again, this will require a broader API fix on PPOCatalog/PPORLModule.